### PR TITLE
fix the incorrect output_tokens for penalty calculation in the sampler when delayed sampling is enabled

### PR DESCRIPTION
**Pull request by ccrhx4**
When delayed sampling is enabled, the output_tokens for penalty calculation in the sampler is incorrect.

The output_tokens should have all the previous output tokens, but when delayed sampling is enabled the last output token id is -1. It had not been updated from the cache.

The fix calls to sync the output_token from cache before sampler is called when penalty is requested in the sampling params.

Please kindly review and give the comment.

### CODE REVIEW

File: vllm/worker/hpu_model_runner.py [hunk 1]
```diff
@@ -2803,7 +2822,8 @@ def try_revert_dummy_output_tokens():
                         output = output.sampled_token_ids
                         self.cached_step_outputs.append(output)
                     if use_delayed_sampling and self.is_driver_worker:
-                        self._patch_prev_output()
+                        if not is_prev_output_patched:
+                            self._patch_prev_output()
```
Comment (madamczyk-intel) [inline vllm/worker/hpu_model_runner.py:2826]: no need to patch here. We can move patching before logits processing (see: https://github.com/HabanaAI/vllm-fork/pull/937/commits/48b140df24e03186365485325e34128a865cb4f3)

**Review by madamczyk-intel — CHANGES_REQUESTED** [review]

File: vllm/worker/hpu_model_runner.py [hunk 2]
```diff
@@ -2784,8 +2786,25 @@ def try_revert_dummy_output_tokens():
...
                     fake_output = self._delayed_sampler_outputs(model_input)
+                    if not model_input.is_prompt:
+                        penalty_are_requested = any([
+                            abs(sg.sampling_params.presence_penalty)
+                            >= _SAMPLING_EPS
+                            or abs(sg.sampling_params.frequency_penalty)
+                            >= _SAMPLING_EPS
+                            or abs(sg.sampling_params.repetition_penalty)
+                            >= _SAMPLING_EPS
+                            for sg in sampling_metadata.seq_groups
```
[the remark this answers is not in the document]
Comment (ccrhx4) [reply vllm/worker/hpu_model_runner.py:2800]: [48b140d](https://github.com/HabanaAI/vllm-fork/commit/48b140df24e03186365485325e34128a865cb4f3) is enough to fix the penalty issue. 
Is it going to be merged into main? If so, my PR will not be necessary. And I prefer the fix in 48b140d.

[the remark this answers is not in the document]
Comment (ccrhx4) [reply vllm/worker/hpu_model_runner.py:2800]: @michalkuligowski You are right, the change is merged. I am closing this PR. Thanks.
